networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationSubnetSubscriptions.java

```diff
@@ -0,0 +1,116 @@
+import io.vertx.core.impl.ConcurrentHashSet;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import tech.pegasys.artemis.networking.eth2.gossip.topics.AttestationTopicHandler;
+import tech.pegasys.artemis.networking.p2p.gossip.GossipNetwork;
+import tech.pegasys.artemis.networking.p2p.gossip.TopicChannel;
+import tech.pegasys.artemis.storage.client.RecentChainData;
+
+public class AttestationSubnetSubscriptions implements AutoCloseable {
+  private final GossipNetwork gossipNetwork;
+  private final RecentChainData recentChainData;
+  private final EventBus eventBus;
+
+  private final Map<Integer, Set<UnsignedLong>> subnetIdToCommittees = new ConcurrentHashMap<>();
+  private final Map<Integer, TopicChannel> subnetIdToTopicChannel = new ConcurrentHashMap<>();
+
+  public AttestationSubnetSubscriptions(
+      final GossipNetwork gossipNetwork,
+      final RecentChainData recentChainData,
+      final EventBus eventBus) {
+    this.gossipNetwork = gossipNetwork;
+    this.recentChainData = recentChainData;
+    this.eventBus = eventBus;
+  }
+
+  public Optional<TopicChannel> getChannel(final UnsignedLong committeeIndex) {
+    final int subnetId = committeeIndexToSubnetId(committeeIndex);
+    if (!subnetIdToCommittees
+        .computeIfAbsent(subnetId, __ -> Collections.emptySet())
+        .contains(committeeIndex)) {
+      // We're not subscribed to this committee subnet
+      return Optional.empty();
+    }
+    return Optional.ofNullable(subnetIdToTopicChannel.get(subnetId));
+  }
+
+  public synchronized void subscribeToCommitteeTopic(final UnsignedLong committeeIndex) {
+    final int subnetId = committeeIndexToSubnetId(committeeIndex);
+    final Set<UnsignedLong> subscribedCommittees =
+        subnetIdToCommittees.computeIfAbsent(subnetId, __ -> new ConcurrentHashSet<>());
+    subscribedCommittees.add(committeeIndex);
+    subnetIdToTopicChannel.computeIfAbsent(subnetId, this::createChannelForSubnetId);
+  }
+
+  public synchronized void unsubscribeFromCommitteeTopic(final UnsignedLong committeeIndex) {
+    final int subnetId = committeeIndexToSubnetId(committeeIndex);
+    final Set<UnsignedLong> subscribedCommittees =
+        subnetIdToCommittees.compute(
+            subnetId,
+            (subnet, committees) -> {
+              if (committees == null) {
+                return null;
+              }
+              committees.remove(committeeIndex);
+              if (committees.isEmpty()) {
+                return null;
+              }
+              return committees;
+            });
+
+    if (subscribedCommittees != null) {
+      // We still have some subscribers, don't actually unsubscribe
+      return;
+    }
+
+    subnetIdToTopicChannel.computeIfPresent(
+        subnetId,
+        (index, channel) -> {
+          channel.close();
+          return null;
+        });
```

Comment:
As an added bonus to using `synchronized` you can then stop using all the concurrency-safe approaches to updating so this method simplifies to something like:
```
final Set<UnsignedLong> committees =
        subnetIdToCommittees.getOrDefault(subnetId, Collections.emptySet());
    committees.remove(committeeIndex);
    if (!committees.isEmpty()) {
      // We still have some subscribers, don't actually unsubscribe
      return;
    }
    subnetIdToCommittees.remove(subnetId);
    final TopicChannel topicChannel = subnetIdToTopicChannel.remove(subnetId);
    topicChannel.close();
```

Reply:
This definitely looks cleaner, although I'm a bit confused about which methods were concurrency-safe whereas the ones you suggested are not.

Reply:
You don't need to write thread-safe code if each public method is synchronised - there will only ever be one thread at a time in the class and memory access between calls is guaranteed to be consistent (ie if call A executes before call B, then call B is guaranteed to see all of call A's changes)

Reply:
Thanks for the explanation. I was also curious what "all the concurrency-safe approaches" in that code were. Because I'm not sure which one of those were concurrency-safe or not.